### PR TITLE
Include build properties in gerrit buildStarted callback

### PR DESCRIPTION
**Pull request by ryansydnor**

### CODE REVIEW

**Comment by tardyp** [comment]
I think we should factorized so that build start and finish callbacks get the same data.

this needs doc and release note, as this is a (backward compatible) API change

**Comment by ryansydnor** [comment]
@tardyp - let me know if this factoring is what you were thinking. I also added release notes and examples of using build properties in the documentation :)

File: master/docs/examples/git_gerrit.cfg [hunk 1]
```diff
@@ -163,7 +166,9 @@ def gerritSummaryCB(buildInfoList, results, status, arg):
             msg += " - " + link
         else:
             msg += "."
-        msgs.append(msg)
+
+        if buildInfo['build']['properties']['shouldReport']:
```
Comment (tardyp) [inline master/docs/examples/git_gerrit.cfg]: shouldReport is a property that you are introducing (that will not exist on others infras)

I would prefer that you dont change the default

File: master/docs/examples/git_gerrit.cfg [hunk 2]
```diff
@@ -144,11 +144,14 @@ def gerritReviewCB(builderName, build, result, master, arg):
     return dict(message=message, labels={'Verified': verified})
 
 def gerritStartCB(builderName, build, arg):
-    print "gerritStartCB..."
+    if not build['properties']['shouldReport']:
+        return dict()
+
     message = "Buildbot started compiling your patchset\n"
     message += "on configuration: %s\n" % builderName
+    message += "See your build here: %s" % build['url']
```
Comment (tardyp) [inline master/docs/examples/git_gerrit.cfg]: build url makes lot of sense  though

File: master/docs/examples/git_gerrit.cfg (hunk 1 above)
Comment (ryansydnor) [inline master/docs/examples/git_gerrit.cfg]: Agreed. I want to demonstrate that the properties are available and can be used. Do you think that's necessary? If so, how would you recommend exemplifying that?

Comment (tardyp) [inline master/docs/examples/git_gerrit.cfg]: You can for sure add this in the documentation

File: master/docs/relnotes/index.rst [hunk 3]
```diff
@@ -36,6 +36,8 @@ Fixes
 
 * OpenStackLatentWorker uses the novaclient API correctly now.
 
+* :bb:reporter:`GerritStatusPush` now includes build properties in the ``startCB`` and ``reviewCB`` functions.
```
Comment (tardyp) [inline master/docs/relnotes/index.rst]: and startCB must  now return a dictionary

Comment (ryansydnor) [inline master/docs/relnotes/index.rst]: Done.

File: master/docs/examples/git_gerrit.cfg [hunk 4]
```diff
@@ -144,11 +144,11 @@ def gerritReviewCB(builderName, build, result, master, arg):
     return dict(message=message, labels={'Verified': verified})
 
 def gerritStartCB(builderName, build, arg):
-    print "gerritStartCB..."
     message = "Buildbot started compiling your patchset\n"
     message += "on configuration: %s\n" % builderName
+    message += "See your build here: %s" % build['url']
 
-    return message
+    return dict(message=message)
```
Comment (sa2ajj) [inline master/docs/examples/git_gerrit.cfg]: [nit] probably `makeReviewResult` would be [a tiny bit] better...

Comment (sa2ajj) [inline master/docs/examples/git_gerrit.cfg]: oh, it's not imported here :/

File: master/buildbot/reporters/gerrit.py [hunk 5]
```diff
@@ -275,32 +275,28 @@ def stopService(self):
...
-
-    @defer.inlineCallbacks
-    def buildFinished(self, builderName, build, result):
-        """Do the SSH gerrit verify command to the server."""
-        result = yield self.reviewCB(builderName, build, result, self.master, self.reviewArg)
-        result = _handleLegacyResult(result)
-        self.sendCodeReviews(build, result)
+        if self.startCB is None:
+            return
+        yield self.getBuildDetails(build)
```
Comment (sa2ajj) [inline master/buildbot/reporters/gerrit.py]: so we run "something" and that "something" has side effects?  if yes, I'd rather not :/

Comment (ryansydnor) [inline master/buildbot/reporters/gerrit.py]: Are you suggesting we remove the functionality (abandon the PR) or we fix the utils.getDetailsForBuilds helper function to operate without side effects?

**Comment by ryansydnor** [comment]
@tardyp - anything else you were looking for here?

**Comment by tardyp** [comment]
we need a rebase :-}

**Pull request merged**